// vitepress-export-pdf/dist/shared/vitepress-export-pdf.514c56e1.cjs
'use strict';

const vuepressPluginExportPdfCore = require('@condorhero/vuepress-plugin-export-pdf-core');
const node_path = require('node:path');
const node_module = require('node:module');
const vitepress = require('vitepress');
const debug = require('debug');
const hash = require('hash-sum');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const debug__default = /*#__PURE__*/_interopDefaultCompat(debug);
const hash__default = /*#__PURE__*/_interopDefaultCompat(hash);

const engines = {
	node: ">=14.6.0"
};
const peerDependencies = {
	vitepress: ">=1.0.0-alpha.35 <=1.x"
};

const devDebug = debug__default("vitepress-export-pdf:dev-server");
const PORT = 16762;
const HOST = "localhost";
const moduleRequire = node_module.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (document.currentScript && document.currentScript.src || new URL('shared/vitepress-export-pdf.514c56e1.cjs', document.baseURI).href)));
const { version } = moduleRequire("vitepress/package.json");
async function serverApp(dir = "docs", commandOptions = {}) {
  vuepressPluginExportPdfCore.checkEnv("VitePress", engines.node, version, peerDependencies.vitepress);
  const sourceDir = node_path.join(process.cwd(), dir);
  if (commandOptions.debug)
    debug__default.enabled("vitepress-export-pdf:*");
  devDebug("sourceDir: %s", sourceDir);
  let userConfig = {};
  const userConfigPath = commandOptions.config ? vuepressPluginExportPdfCore.resolveUserConfigPath(commandOptions.config) : vuepressPluginExportPdfCore.resolveUserConfigConventionalPath(sourceDir, "vitepress");
  if (userConfigPath)
    userConfig = await vuepressPluginExportPdfCore.loadModule(userConfigPath);
  if (Array.isArray(userConfig.routePatterns))
    userConfig.routePatterns = ["/**", "!/404.html", ...userConfig.routePatterns];
  else
    userConfig.routePatterns = ["/**", "!/404.html"];
  const vitepressOutFile = commandOptions.outFile ?? `vitepress-${vuepressPluginExportPdfCore.timeTransformer()}.pdf`;
  const vitepressOutDir = commandOptions.outDir ?? ".";
  devDebug("userConfig: %O", userConfig);
  const {
    sorter,
    puppeteerLaunchOptions,
    pdfOptions,
    routePatterns,
    outFile = vitepressOutFile,
    outDir = vitepressOutDir,
    urlOrigin = commandOptions.urlOrigin,
    pdfOutlines = commandOptions.pdfOutlines,
    outlineContainerSelector = ".VPContent"
  } = userConfig;
  const devServer = await vitepress.createServer(sourceDir, { port: PORT, host: HOST });
  const { port = PORT, host = HOST } = devServer.config.server;
  const devApp = await devServer.listen(port);
  devApp.printUrls();
  process.stdout.write("\n");
  process.stdout.write("Start to generate current site to PDF ...");
  const { pages, tempDir, cleanUrls } = await vitepress.resolveConfig(devApp.config.root);
  const haveCleanUrls = cleanUrls ? "" : ".html";
  const hashPages = pages.map((page) => ({
    path: `${devServer.config.base}${page.replace(/\.md$/, haveCleanUrls)}`,
    key: `v-${hash__default(page)}`
  }));
  try {
    await vuepressPluginExportPdfCore.generatePdf({
      pages: hashPages,
      tempDir,
      host: typeof host === "boolean" ? HOST : host,
      port,
      outFile,
      outDir,
      sorter,
      urlOrigin,
      pdfOptions,
      pdfOutlines,
      routePatterns,
      puppeteerLaunchOptions,
      outlineContainerSelector
    });
  } catch (error) {
    console.error(error);
  }
  await devApp.close();
  process.exit(0);
}

function registerCommands(program) {
  program.command("export [sourceDir]", "Export current VitePress site to a PDF file(default: docs)").allowUnknownOptions().option("-c, --config <config>", "Set path to config file").option("--outFile <outFile>", "Name of output file").option("--outDir <outDir>", "Directory of output files").option("--pdfOutlines <pdfOutlines>", "Keep PDF outlines/bookmarks").option("--urlOrigin <urlOrigin>", "Change the origin of the print url(Option displayHeaderFooter of pdfOptions is true)").option("--debug", "Enable debug mode").action(serverApp);
  program.command("info", "Display environment information").action(() => {
    vuepressPluginExportPdfCore.systemInfo(["vitepress"]);
  });
}
vuepressPluginExportPdfCore.runCli("press-export-pdf")(registerCommands);

exports.HOST = HOST;
exports.PORT = PORT;
exports.moduleRequire = moduleRequire;
exports.registerCommands = registerCommands;
exports.serverApp = serverApp;
