// @condorhero/vuepress-plugin-export-pdf-core/dist/runner.mjs
import cac from 'cac';

const name = "@condorhero/vuepress-plugin-export-pdf-core";
const type = "module";
const version = "3.0.2";
const description = "The Core of VuePress and VitePress exports PDF plugin";
const author = "CondorHero";
const license = "MIT";
const homepage = "https://github.com/condorheroblog/vuepress-plugin-export-pdf#readme";
const repository = {
	type: "git",
	url: "git+https://github.com/condorheroblog/vuepress-plugin-export-pdf.git",
	directory: "packages/vuepress-plugin-export-pdf-core"
};
const bugs = {
	url: "https://github.com/condorheroblog/vuepress-plugin-export-pdf/issues"
};
const keywords = [
	"vuepress-plugin-export-pdf",
	"vuepress-plugin",
	"vuepress",
	"plugin"
];
const exports = {
	".": {
		types: "./dist/index.d.ts",
		require: "./dist/index.cjs",
		"import": "./dist/index.mjs"
	}
};
const main = "dist/index.cjs";
const module = "dist/index.mjs";
const types = "dist/index.d.ts";
const typesVersions = {
	"*": {
		"*": [
			"./dist/*",
			"./dist/index.d.ts"
		]
	}
};
const files = [
	"dist"
];
const engines = {
	node: ">=8.6"
};
const scripts = {
	build: "unbuild",
	prepublishOnly: "nr build",
	dev: "esno ./src/index.ts"
};
const dependencies = {
	"@condorhero/merge-pdfs": "^0.0.7",
	"bundle-require": "^4.0.1",
	cac: "^6.7.14",
	envinfo: "^7.10.0",
	"fs-extra": "^11.1.1",
	"html-export-pdf-cli": "^1.1.4",
	multimatch: "^6.0.0",
	ora: "^6.3.1",
	pdfjs: "^2.5.0",
	picocolors: "^1.0.0",
	semver: "^7.5.3"
};
const devDependencies = {
	"@types/envinfo": "^7.8.1",
	"@types/fs-extra": "^11.0.1",
	"@types/semver": "^7.5.0"
};
const publishConfig = {
	access: "public",
	registry: "https://registry.npmjs.org"
};
const pkg = {
	name: name,
	type: type,
	version: version,
	description: description,
	author: author,
	license: license,
	homepage: homepage,
	repository: repository,
	bugs: bugs,
	keywords: keywords,
	exports: exports,
	main: main,
	module: module,
	types: types,
	typesVersions: typesVersions,
	files: files,
	engines: engines,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies,
	publishConfig: publishConfig
};

const beforeParse = (cliInstance) => {
  cliInstance.version(pkg.version).help();
};
const afterParse = (cliInstance) => {
  if (!process.argv.slice(2).filter(Boolean).length)
    cliInstance.outputHelp();
};
const runCli = (programName) => {
  try {
    const program = cac(programName);
    return (registerCommands) => {
      registerCommands(program);
      beforeParse(program);
      program.parse(process.argv);
      afterParse(program);
    };
  } catch (error) {
    process.exit(1);
  }
};

export { afterParse, beforeParse, runCli };
