// @condorhero/vuepress-plugin-export-pdf-core/dist/systemInfo.cjs
'use strict';

const pc = require('picocolors');
const envinfo = require('envinfo');
const ora = require('ora');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const pc__default = /*#__PURE__*/_interopDefaultCompat(pc);
const envinfo__default = /*#__PURE__*/_interopDefaultCompat(envinfo);
const ora__default = /*#__PURE__*/_interopDefaultCompat(ora);

const systemInfo = async (message) => {
  const spinner = ora__default();
  spinner.start(pc__default.bold("\nCollecting Environment Info:"));
  const result = await envinfo__default.run(
    {
      System: ["OS", "CPU", "Memory", "Shell"],
      Binaries: ["Node", "Yarn", "npm"],
      Utilities: ["Git"],
      Browsers: ["Chrome", "Edge", "Firefox", "Safari"],
      npmPackages: message
    },
    {
      showNotFound: true,
      duplicates: true,
      fullTree: true
    }
  );
  spinner.stop();
  process.stdout.write(result);
  return result;
};

exports.systemInfo = systemInfo;
